torch/quantization/_adaround.py

```diff
@@ -0,0 +1,260 @@
+    weights_clipped = weights_divided_by_scale + clipped_sigmoid(continous_V)
+
+    weights_w_adaround_rounding = scale * torch.clamp(weights_clipped, model.quant_min, model.quant_max)
+    return weights_w_adaround_rounding
+
+def loss_function_leaf(model, count):
+    ''' Calculates the loss function for a submodule
+
+    Paper Reference: https://arxiv.org/pdf/2004.10568.pdf Eq. 25
+    '''
+    beta = count / number_of_epochs * (beta_high - beta_low) + beta_low
+
+
+    # Calculates the difference between floating point and quantized models
+    adaround_instance = model.wrapped_module.weight_fake_quant
+    float_weight = model.wrapped_module.weight
+    clipped_weight = adaround_rounding(adaround_instance, float_weight)
+    quantized_weight = torch.fake_quantize_per_tensor_affine(clipped_weight, float(adaround_instance.scale),
+                                                             int(adaround_instance.zero_point), adaround_instance.quant_min,
+                                                             adaround_instance.quant_max)
+    # Frobenius_norm = torch.norm(float_weight - quantized_weight)  # norm(W - ~W)
+    Frobenius_norm = torch.norm(model.float_output - model.quantized_output)  # norm(Wx - ~Wx)
+
+    # calculating regularization factor -> forces values of continous_V to be 0 or 1
+    scale = adaround_instance.scale
+    continous_V = adaround_instance.continous_V
+    clip_V = clipped_sigmoid(continous_V)
+    spreading_range = torch.abs((2 * clip_V) - 1)
+    one_minus_beta = 1 - (spreading_range ** beta)  # torch.exp
+    regulization = torch.sum(one_minus_beta)
+
+    print("loss function break down: ", Frobenius_norm * norm_scaling, regularization_scaling * regulization)
+    # print("sqnr of float and quantized: ", computeSqnr(float_weight, quantized_weight))
+    return Frobenius_norm * norm_scaling + regularization_scaling * regulization
+
+def loss_function(model, count):
+    ''' Given model, the loss function of all of its whitelisted leaf modules will
+    be added up and returned.
+    '''
+    result = torch.Tensor([0])
+    for name, submodule in model.named_modules():
+        if isinstance(submodule, OuputWrapper):
+            result = result + loss_function_leaf(submodule, count)
+            result = result * .90
+    return result
+
+def add_wrapper_class(model, white_list=DEFAULT_QAT_MODULE_MAPPING.keys()):
+    ''' Throws on a wrapper class to collect the last output passed through it.
+    This information is used in computing the loss function.
+    '''
+    for name, submodule in model.named_modules():
+        if type(submodule) in white_list:
+            parent_name, child_name = parent_child_names(name)
+            parent_module = get_module(model, parent_name)
+            parent_module._modules[child_name] = OuputWrapper(submodule)
+            submodule.qconfig = adaround_qconfig
+
+
+def prepare_adaround(float_model, dataset, white_list=_supported_modules):
```

Comment:
users are supposed to attach qconfig themselves in eager mode, I don't think we should call prepare_qat for user. AdaRoundFakeQuantize module will just be a module that user can use in replacement of FakeQuantize module. we should implement loss functions etc. as a method of AdaRoundFakeQuantize module itself.